File: dmux_verify/report.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .guide import load_stages
from .state import ProgressStore


def markdown_report(stages, store: ProgressStore) -> str:
    lines = ["# DurableMux Verification Report", ""]
    passed = sum(1 for n in stages if store.is_passed(n))
    lines.append(f"**Progress:** {passed}/{len(stages)} stages passed")
    lines.append("")
    lines.append("| Stage | Title | Mode | Status | Evidence |")
    lines.append("|---:|---|---|---|---:|")
    for number, stage in stages.items():
        item = store.stage(number)
        status = item.get("status", "locked" if number > 1 and not store.is_passed(number - 1) else "ready")
        lines.append(f"| {number} | {stage.title} | {stage.mode} | {status} | {len(item.get('evidence', []))} |")
    lines.append("")
    for number, stage in stages.items():
        item = store.stage(number)
        if not item.get("last_run") and not item.get("evidence"):
            continue
        lines.extend([f"## Stage {number}: {stage.title}", ""])
        run = item.get("last_run")
        if run:
            lines.append(f"Result: **{'PASS' if run.get('passed') else 'FAIL'}**")
            lines.append("")
            for check in run.get("checks", []):
                icon = "PASS" if check.get("passed") else "FAIL"
                lines.append(f"- **{icon}:** {check.get('name')} — {check.get('detail', '')}")
        evidence = item.get("evidence", [])
        if evidence:
            lines.append("")
            lines.append("Evidence:")
            for entry in evidence:
                lines.append(f"- {entry.get('path') or entry.get('command')} — {entry.get('note', '')}")
        lines.append("")
    return "\n".join(lines)

File: internal/report/report.go
package report

import (
	"fmt"
	"strings"

	"durablemux-verifier/internal/model"
	"durablemux-verifier/internal/store"
)

// Markdown builds the progress report (parity with Python report.py).
func Markdown(stages map[int]*model.Stage, s *store.Store) string {
	var b strings.Builder
	b.WriteString("# DurableMux Verification Report\n\n")
	passed := 0
	for n := range stages {
		if s.IsPassed(n) {
			passed++
		}
	}
	fmt.Fprintf(&b, "**Progress:** %d/%d stages passed\n\n", passed, len(stages))
	b.WriteString("| Stage | Title | Mode | Status | Evidence |\n")
	b.WriteString("|---:|---|---|---|---:|\n")
	for _, n := range sortedKeys(stages) {
		st := stages[n]
		item := s.Stage(n)
		status, _ := item["status"].(string)
		if status == "" {
			if n > 1 && !s.IsPassed(n-1) {
				status = "locked"
			} else {
				status = "ready"
			}
		}
		ev, _ := item["evidence"].([]any)
		fmt.Fprintf(&b, "| %d | %s | %s | %s | %d |\n", n, st.Title, st.Mode, status, len(ev))
	}
	b.WriteString("\n")
	for _, n := range sortedKeys(stages) {
		st := stages[n]
		item := s.Stage(n)
		if item["last_run"] == nil && item["evidence"] == nil {
			continue
		}
		fmt.Fprintf(&b, "## Stage %d: %s\n\n", n, st.Title)
		if run, ok := item["last_run"].(map[string]any); ok {
			if run["passed"] == true {
				b.WriteString("Result: **PASS**\n\n")
			} else {
				b.WriteString("Result: **FAIL**\n\n")
			}
			if checks, ok := run["checks"].([]any); ok {
				for _, c := range checks {
					if m, ok := c.(map[string]any); ok {
						icon := "FAIL"
						if m["passed"] == true {
							icon = "PASS"
						}
						fmt.Fprintf(&b, "- **%s:** %v — %v\n", icon, m["name"], m["detail"])
					}
				}
			}
		}
		if ev, ok := item["evidence"].([]any); ok && len(ev) > 0 {
			b.WriteString("\nEvidence:\n")
			for _, e := range ev {
				if m, ok := e.(map[string]any); ok {
					ref := m["path"]
					if ref == nil {
						ref = m["command"]
					}
					fmt.Fprintf(&b, "- %v — %v\n", ref, m["note"])
				}
			}
		}
		b.WriteString("\n")
	}
	return b.String()
}

func sortedKeys(stages map[int]*model.Stage) []int {
	keys := make([]int, 0, len(stages))
	for k := range stages {
		keys = append(keys, k)
	}
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j] < keys[j-1]; j-- {
			keys[j], keys[j-1] = keys[j-1], keys[j]
		}
	}
	return keys
}
